### PR TITLE
Add ConvNeXt Mask R-CNN

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR is an initial draft for implementing the classic Mask R-CNN framework with ConvNeXt as backbone. 

The framework is implemented in a single script, with the exception of 3 files (for now):
* assign_result.py
* losses.py
* mask_target.py

As we have a one model, one file policy, I'm reimplementing ConvNeXT leveraging Copied from statements. So `ConvNextMaskRCNNModel` is almost identical to `ConvNextModel`. This way, the backbone used for object detection stays independent from the original one. In this case for instance, extra layernorms are added after each stage.

There's a dependency on torchvision, which is used for NMS (non-maximum suppression, a postprocessing algorithm used by both the RPN head and the RoI head).

To do:

- [x] update NumPy logic to pure PyTorch (i.e. channels first everywhere) - see branch `add_convnext_maskrcnn_torch_shapes`
- [ ] update outputs of model to have channels first (no NumPy)

### CODE REVIEW

File: src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py [hunk 1]
```diff
@@ -0,0 +1,3526 @@
...
+        bbox_weights_list = images_to_levels(all_bbox_weights, num_level_anchors)
+        res = (labels_list, label_weights_list, bbox_targets_list, bbox_weights_list, num_total_pos, num_total_neg)
+        if return_sampling_results:
+            res = res + (sampling_results_list,)
+        for i, r in enumerate(rest_results):  # user-added return values
+            rest_results[i] = images_to_levels(r, num_level_anchors)
+
+        return res + tuple(rest_results)
+
+    def loss_single(
```
Comment (amyeroberts) [inline src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py:2073]: Could you rename it to make clearer what it does? e.g. 
```suggestion
    def loss_single_scale_level(
```

File: src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py [hunk 2]
```diff
@@ -0,0 +1,3526 @@
...
+        losses_cls, losses_bbox = multi_apply(
+            self.loss_single,
+            cls_scores,
+            bbox_preds,
+            all_anchor_list,
+            labels_list,
+            label_weights_list,
+            bbox_targets_list,
+            bbox_weights_list,
+            num_total_samples=num_total_samples,
```
Comment (amyeroberts) [inline src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py:2184]: I'm a bit confused about `num_total_samples` here - it might just be my understanding. Following the code, I see it's used as `avg_factor` in the loss calculation when calculating the mean loss. As it's being passed into `loss_single` I would have expected this to be the total number of samples for that scale level. Looking at `get_targets` it looks like it's the sum across all levels i.e. if another level was added then the losses would go down.

File: src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py [hunk 3]
```diff
@@ -0,0 +1,3526 @@
...
+            a kind of returned results by the function
+    """
+    pfunc = partial(func, **kwargs) if kwargs else func
+    map_results = map(pfunc, *args)
+    return tuple(map(list, zip(*map_results)))
+
+
+def unmap(data, count, inds, fill=0):
+    """Unmap a subset of item (data) back to the original set of items (of size
+    count)"""
```
Comment (amyeroberts) [inline src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py:122]: Could you fill out this function a bit - types, docstring etc. It's not clear to me what is meant by "unmap" or the intended behaviour.

**Review by amyeroberts — COMMENTED** [review]
Nice! Looking good. Some parts I didn't review deeply because it was obviously still WIP e.g. `losses.py` module. Happy to look at more if you want. Overall structure and code looks good 👍

You're probably already on it, but it would be great to see some tests for the NMS logic. 

Main comment is about `multi_apply` and how things are called. I don't see the reason for having this function - I might be missing something - but a simple for loop would be easier to understand and wouldn't need the additional logic to create lists to pass into `multi_apply` e.g. L2019-2032 in `modeling_convnext_maskrcnn.py`. 

I also think a lot of this logic could be simplified if most of the methods were re-written to operate on single elements / levels. For example, we could write `get_anchor`, `valid_flag`, `get_target` instead of the plurarised versions. Then `loss` could have a simple for-loop. This would remove the need for logic to handle lists in so many places.

File: utils/check_repo.py [hunk 4]
```diff
@@ -125,10 +125,10 @@
 # should **not** be the rule.
 IGNORE_NON_AUTO_CONFIGURED = PRIVATE_MODELS.copy() + [
     # models to ignore for model xxx mapping
+    "ConvNextMaskRCNNForObjectDetection",
     "DPTForDepthEstimation",
     "DecisionTransformerGPT2Model",
     "GLPNForDepthEstimation",
-    "ViltForQuestionAnswering",
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply utils/check_repo.py:131]: I'll address this in a separate PR

File: src/transformers/models/convnext/configuration_convnext.py [hunk 5]
```diff
@@ -88,7 +88,6 @@ def __init__(
         hidden_act="gelu",
         initializer_range=0.02,
         layer_norm_eps=1e-12,
-        is_encoder_decoder=False,
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/convnext/configuration_convnext.py:91]: Because it's False by default

File: src/transformers/models/convnext_maskrcnn/convert_convnext_maskrcnn_to_pytorch.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+logger = logging.get_logger(__name__)
+
+
+def get_convnext_maskrcnn_config():
+    config = ConvNextMaskRCNNConfig()
+
+    config.num_labels = 80
+    repo_id = "datasets/huggingface/label-files"
+    filename = "coco-detection-mmdet-id2label.json"
+    id2label = json.load(open(hf_hub_download(repo_id, filename), "r"))
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/convnext_maskrcnn/convert_convnext_maskrcnn_to_pytorch.py:45]: What's the enefit of a context manager? Python just recommends using json.load: https://docs.python.org/3/library/json.html

File: src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py [hunk 7]
```diff
@@ -0,0 +1,3526 @@
...
+        return input
+    keep_prob = 1 - drop_prob
+    shape = (input.shape[0],) + (1,) * (input.ndim - 1)  # work with diff dim tensors, not just 2D ConvNets
+    random_tensor = keep_prob + torch.rand(shape, dtype=input.dtype, device=input.device)
+    random_tensor.floor_()  # binarize
+    output = input.div(keep_prob) * random_tensor
+    return output
+
+
+# Copied from transformers.models.beit.modeling_beit.BeitDropPath with Beit->ConvNextMaskRCNN
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py:675]: Good point, will update

File: src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py [hunk 8]
```diff
@@ -0,0 +1,3526 @@
...
+
+    def forward(self, hidden_states: torch.FloatTensor) -> torch.Tensor:
+        hidden_states = self.downsampling_layer(hidden_states)
+        hidden_states = self.layers(hidden_states)
+        return hidden_states
+
+
+class ConvNextMaskRCNNEncoder(nn.Module):
+    """
+    This class isn't copied from modeling_convnext.py as layernorms are added.
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py:822]: Ok!

File: src/transformers/models/convnext_maskrcnn/convert_convnext_maskrcnn_to_pytorch.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+    for label in range(len(bbox_results)):
+        if len(bbox_results[label]) > 0:
+            for detection in bbox_results[label]:
+                detections.append((label, detection))
+
+    predicted_classes = [det[0] for det in detections]
+    assert predicted_classes == [15, 15, 57, 57, 59, 65, 65, 79]
+    predicted_bbox = detections[0][1]
+    assert np.allclose(
+        predicted_bbox, np.array([17.905708, 55.41647, 318.95575, 470.25925, 0.9981325], dtype=np.float32), atol=10
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/convnext_maskrcnn/convert_convnext_maskrcnn_to_pytorch.py:162]: No, I had an issue there, will update

File: src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py [hunk 10]
```diff
@@ -0,0 +1,3526 @@
...
+        flat_anchors (torch.Tensor): Flatten anchors, shape (n, 4).
+        valid_flags (torch.Tensor): An existing valid flags of anchors.
+        img_shape (tuple(int)): Shape of current image.
+        allowed_border (int, optional): The border to allow the valid anchor.
+            Defaults to 0.
+    Returns:
+        torch.Tensor: Flags indicating whether the anchors are inside a \
+            valid range.
+    """
+    img_h, img_w = img_shape[:2]
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py:186]: Yes, they seem to rely on NumPy a lot

**Review by alaradirik — COMMENTED** [review]
The draft looks good to me but requires a lot of restructuring. I noticed that a lot of the loss metrics and utility functions could be used by other models or by users during training. 

I'd do a more through pass over `modeling_convnext_maskrcnn.py` after the current issues are addressed.

Good work! :)

File: src/transformers/models/convnext_maskrcnn/convert_convnext_maskrcnn_to_pytorch.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/convnext_maskrcnn/convert_convnext_maskrcnn_to_pytorch.py:45]: I was recommending the context manager for the use of `open`, not `json.load`. In the linked python docs, the `fp` passed to `json.load` is a file-like object created using `StringIO`. 

As @alaradirik says, the CM helps safely open, read and then close a file.

File: src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py [hunk 11]
```diff
@@ -0,0 +1,3525 @@
...
+        boxes = torch.from_numpy(boxes)
+    if isinstance(scores, np.ndarray):
+        scores = torch.from_numpy(scores)
+    assert boxes.size(1) == 4
+    assert boxes.size(0) == scores.size(0)
+    assert offset in (0, 1)
+
+    # TODO use the NMS op of mmcv: https://github.com/open-mmlab/mmcv/blob/d71d067da19d71d79e7b4d7ae967891c7bb00c05/mmcv/ops/nms.py#L28
+    # inds = NMSop.apply(boxes, scores, iou_threshold, offset, score_threshold, max_num)
+    inds = torchvision.ops.nms(boxes, scores, iou_threshold)
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py:370]: I've explicitly replaced the NMS op implemented in mmcv by torchvision.

I think we'll need the torchvision dependency for this model, since they provide a well-known NMS implementation.

File: src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py [hunk 12]
```diff
@@ -0,0 +1,3525 @@
...
+
+
+def nms(
+    boxes: ArrayType,
+    scores: ArrayType,
+    iou_threshold: float,
+    offset: int = 0,
+    score_threshold: float = 0,
+    max_num: int = -1,
+) -> Tuple[ArrayType, ArrayType]:
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py:338]: Yes we probably need to do that!

File: src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py [hunk 13]
```diff
@@ -0,0 +1,3526 @@
...
+        3
+    """
+    assert isinstance(boxes, (torch.Tensor, np.ndarray))
+    assert isinstance(scores, (torch.Tensor, np.ndarray))
+    is_numpy = False
+    if isinstance(boxes, np.ndarray):
+        is_numpy = True
+        boxes = torch.from_numpy(boxes)
+    if isinstance(scores, np.ndarray):
+        scores = torch.from_numpy(scores)
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py:367]: Ah, OK. I didn't realise that. In that case, I wouldn't add anything as I don't think this should be an exception.

[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/convnext_maskrcnn/modeling_convnext_maskrcnn.py:367]: Ok, closing this in that case
